File: ReactiveStreams/src/ReactiveTweets/Reactive.Tweets/Data/EmotionAnalysis.cs
using edu.stanford.nlp.ling;
using edu.stanford.nlp.neural.rnn;
using edu.stanford.nlp.pipeline;
using edu.stanford.nlp.sentiment;
using edu.stanford.nlp.trees;
using java.util;
using System;
using System.Linq;
using Tweetinvi.Models;

namespace Reactive.Tweets
{
    public class EmotionAnalysis
    {
        public EmotionAnalysis(string modelsFolder)
        {
            var properties = new Properties();
            properties.setProperty("annotators",
                "tokenize,ssplit,pos,parse,sentiment");

            var temp = Environment.CurrentDirectory;
            Environment.CurrentDirectory = modelsFolder;
            stanfordNLP = new StanfordCoreNLP(properties);
            Environment.CurrentDirectory = temp;

            emotionAnnotationTreeClassName =
                new SentimentCoreAnnotations.SentimentAnnotatedTree().getClass();
            sentencesAnnotationClassName =
                new CoreAnnotations.SentencesAnnotation().getClass();
        }

        private readonly StanfordCoreNLP stanfordNLP;
        private readonly java.lang.Class emotionAnnotationTreeClassName;
        private readonly java.lang.Class sentencesAnnotationClassName;

        public class Emotion
        {
            public Emotion(int num)
            {
                _num = num;
            }
            private readonly int _num;
            public override string ToString()
            {
                switch (_num)
                {
                    case 0: case 1:
                        return "Unhappy";
                    case 2:
                        return "Indifferent";
                    case 3: case 4:
                        return "Happy";
                    default:
                        throw new ArgumentException($"Unknown emotion value '{_num}'");
                }
            }
        }

        public Emotion GetEmotion(string text)
        {
            var annotation = new Annotation(text);
            stanfordNLP.annotate(annotation);

            return
                (annotation.get(sentencesAnnotationClassName) as ArrayList)
                .toArray().Select(ParseEmotion).FirstOrDefault();

            Emotion ParseEmotion(object s) {
                var sentence = s as Annotation;
                var sentenceTree = sentence.get(emotionAnnotationTreeClassName) as Tree;
                var emotion = RNNCoreAnnotations.getPredictedClass(sentenceTree);
                return new Emotion(emotion);
            }
        }

        public (ITweet, Emotion) AddEmotion(ITweet tweet)
        {
            var emotion = GetEmotion(tweet.Text);
            return (tweet, emotion);
        }
    }
}

File: ReactiveStreams/src/ReactiveTweets/Reactive.Stream/EmotionAnalysis.fs
namespace Reactive.Emotion

type EmotionType = { emotion:int }
                   override this.ToString() =
                      match this.emotion with
                      | 0 | 1 -> "Unhappy"
                      | 2 -> "Indifferent"
                      | 3 | 4 -> "Happy"
                      | x -> failwith (sprintf "Unknown emotion value %d" x)

module Analysis =

    open edu.stanford.nlp.ling
    open edu.stanford.nlp.neural.rnn
    open edu.stanford.nlp.pipeline
    open edu.stanford.nlp.sentiment
    open edu.stanford.nlp.trees
    open java.util
    open System
    open System.Linq
    open Tweetinvi.Models

    let modelsFolder = "../../../models"
    let properties = Properties()
    properties.setProperty("annotators", "tokenize,ssplit,pos,parse,sentiment") |> ignore

    let temp = Environment.CurrentDirectory
    Environment.CurrentDirectory <- modelsFolder
    let stanfordNLP = new StanfordCoreNLP(properties)
    Environment.CurrentDirectory <- temp

    let emotionAnnotationTreeClassName = 
        let sTree = new SentimentCoreAnnotations.SentimentAnnotatedTree()
        sTree.getClass()

    let sentencesAnnotationClassName = 
        let sentence = new CoreAnnotations.SentencesAnnotation()
        sentence.getClass()

    let parseEmotion (o:obj) =
        let sentence = o :?> Annotation
        let sentenceTree = sentence.get(emotionAnnotationTreeClassName) :?> Tree
        let emotion = RNNCoreAnnotations.getPredictedClass(sentenceTree)
        { emotion = emotion } 

    let getEmotion (text:string) =
        let annotation = new Annotation(text)
        stanfordNLP.annotate(annotation)
        let arr = annotation.get(sentencesAnnotationClassName) :?> ArrayList
        arr.toArray()
        |> Array.map(fun o -> parseEmotion o)
        |> Array.head
    
    let addEmotion (tweet:ITweet) =
        let emotion = getEmotion tweet.Text
        (tweet, emotion)
